data/transactions/logic/pairing.go

```diff
@@ -17,99 +17,876 @@
+	points, err := bytesToBLS12381G2s(pointBytes, false)
+	if err != nil {
+		return nil, err
+	}
+	if len(scalarBytes) != scalarSize*len(points) {
+		return nil, fmt.Errorf("bad scalars length %d. Expected %d", len(scalarBytes), scalarSize*len(points))
+	}
+	if len(points) <= bls12381G2MultiExpThreshold {
+		return bls12381G2MultiExpSmall(points, scalarBytes)
+	}
+	return bls12381G2MultiExpLarge(points, scalarBytes)
+}
+
+func bls12381G2MultiExpLarge(points []bls12381.G2Affine, scalarBytes []byte) ([]byte, error) {
+	scalars := make([]bls12381fr.Element, len(points))
+	for i := range scalars {
+		scalars[i].SetBytes(scalarBytes[i*scalarSize : (i+1)*scalarSize])
+	}
+	res, err := new(bls12381.G2Affine).MultiExp(points, scalars, eccMontgomery)
+	if err != nil {
+		return nil, err
+	}
+	return bls12381G2ToBytes(res), nil
+}
+
+func bls12381G2MultiExpSmall(points []bls12381.G2Affine, scalarBytes []byte) ([]byte, error) {
+	// There must be at least one point. Start with it, rather than the identity.
+	k := new(big.Int).SetBytes(scalarBytes[:scalarSize])
+	var sum bls12381.G2Jac
+	sum.FromAffine(&points[0])
+	sum.ScalarMultiplication(&sum, k)
+	for i := range points {
+		if i == 0 {
+			continue
+		}
+		k.SetBytes(scalarBytes[i*scalarSize : (i+1)*scalarSize])
+		var prod bls12381.G2Jac
+		prod.FromAffine(&points[i])
+		prod.ScalarMultiplication(&prod, k)
+		sum.AddAssign(&prod)
+	}
+	var res bls12381.G2Affine
+	res.FromJacobian(&sum)
+	return bls12381G2ToBytes(&res), nil
+}
+
+func bls12381MapToG1(fpBytes []byte) ([]byte, error) {
+	fp, err := bytesToBLS12381Field(fpBytes)
+	if err != nil {
+		return nil, err
+	}
+	point := bls12381.MapToG1(fp)
+	return bls12381G1ToBytes(&point), nil
+}
+
+func bls12381MapToG2(fpBytes []byte) ([]byte, error) {
+	if len(fpBytes) != bls12381fp2Size {
+		return nil, fmt.Errorf("bad encoded element length: %d", len(fpBytes))
+	}
+	g2 := bls12381.G2Affine{}
+	var err error
+	g2.X.A0, err = bytesToBLS12381Field(fpBytes[0:bls12381fpSize])
+	if err != nil {
+		return nil, err
+	}
+	g2.X.A1, err = bytesToBLS12381Field(fpBytes[bls12381fpSize:])
+	if err != nil {
+		return nil, err
+	}
+	point := bls12381.MapToG2(g2.X)
+	return bls12381G2ToBytes(&point), nil
+}
+
+func bls12381G1SubgroupCheck(pointBytes []byte) (bool, error) {
+	point, err := bytesToBLS12381G1(pointBytes)
+	if err != nil {
+		return false, err
+	}
+	return point.IsInSubGroup(), nil
+}
+
+func bls12381G2SubgroupCheck(pointBytes []byte) (bool, error) {
+	point, err := bytesToBLS12381G2(pointBytes)
+	if err != nil {
+		return false, err
+	}
+	return point.IsInSubGroup(), nil
+}
+
+var bn254Modulus = bn254fp.Modulus()
+
+func bytesToBN254Field(b []byte) (bn254fp.Element, error) {
+	var big big.Int
+	big.SetBytes(b)
+	if big.Cmp(bn254Modulus) >= 0 {
+		return bn254fp.Element{}, fmt.Errorf("field element %s larger than modulus %s", &big, bn254Modulus)
+	}
+	return *new(bn254fp.Element).SetBigInt(&big), nil
```

Comment:
nit: if not compiled away, this also seems like an unnecessary use of `new`

Reply:
I tried this one too, since less was going on here, I had hopes that a stack variable would not escape. But it didn't so allocs unchanged.